lxd/container.go

```diff
@@ -214,118 +172,19 @@ func containerValidConfig(sysOS *sys.OS, config map[string]string, profile bool,
 	return nil
 }
 
-func containerValidDevices(state *state.State, cluster *db.Cluster, devices config.Devices, profile bool, expanded bool) error {
-	// Empty device list
-	if devices == nil {
-		return nil
-	}
-
-	var diskDevicePaths []string
-	// Check each device individually
-	for name, m := range devices {
-		// Validate config using device interface.
-		_, err := device.New(&containerLXC{}, state, name, config.Device(m), nil, nil)
-		if err != device.ErrUnsupportedDevType {
-			if err != nil {
-				return err
-			}
-			continue // Validated device OK.
-		}
-
-		// Fallback to legacy validation for non device package devices.
-		if !shared.StringInSlice(m["type"], []string{"disk", "none"}) {
-			return fmt.Errorf("Invalid device type for device '%s'", name)
-		}
-
-		for k := range m {
-			if !containerValidDeviceConfigKey(m["type"], k) {
-				return fmt.Errorf("Invalid device configuration key for %s: %s", m["type"], k)
-			}
+// containerValidDevices validate container device configs.
+func containerValidDevices(state *state.State, cluster *db.Cluster, instance device.InstanceIdentifier, devices deviceConfig.Devices, expanded bool) error {
```

Comment:
@stgraber I've removed the `profile` boolean argument from this function, as each device can use the presence of an empty instance name to indicate profile validation mode (if needed).

However I am unclear on the role of the `expanded` boolean in this function, and how it relates to disk validation. As it is, the new code doesn't take this into account and proceeds with normal validation irrespective of this boolean value, so I would like to remove it ideally, but I am unclear what it is for.

Reply:
I believe we sorted that part out on IRC, right?

Reply:
@stgraber yes thanks, I've implemented that now. I've still kept the `expanded` boolean on this function and use that to indicate how to populate a temporary `containerLXC` struct such that the population of `expandedDevices` indicates an expanded validation.